url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/utils/Async.java

```diff
@@ -0,0 +1,283 @@
+        protected T object;
+
+        /**
+         * Returns the object payload.
+         * 
+         * @return the object The object payload.
+         */
+        public T getObject() {
+            return this.object;
+        }
+
+        protected QEntry() {}
+    }
+
+    /** Represents the end of queue, after this nothing will be sent anymore */
+    protected static class EOQ extends QEntry<Object> {}
+
+    /** The actual EOQ object */
+    protected final static EOQ EOQ = new EOQ();
+
+    /**
+     * The queue feeding the {@link Async} object.
+     * 
+     * @author Ralf Biedert
+     * @param <T> the type of object.
+     * @since 1.0
+     */
+    public static class Queue<T> extends LinkedBlockingQueue<QEntry<T>> {
+        /** */
+        private static final long serialVersionUID = -5727467218361481807L;
+
+        protected Queue() {}
+
+        /**
+         * Closes the queue and signals the receiver that nothing more will be added.
+         * 
+         * @since 1.0
+         */
+        @SuppressWarnings("unchecked")
+        public void close() {
+            add((net.jcores.jre.utils.Async.QEntry<T>) EOQ);
+        }
+    }
+
+    /**
+     * Constructs a communications queue.
+     * 
+     * @since 1.0
+     * @return Returns the queue.
+     */
+    public static final <T> Queue<T> Queue() {
+        return new Queue<T>();
+    }
+
+    /**
+     * Creates a queue entry for the given object.
+     * 
+     * @since 1.0
+     * @param toWrap The object to wrap.
+     * @return The generated queue entry.
+     */
+    public static final <T> QEntry<T> QEntry(T toWrap) {
+        QEntry<T> qe = new QEntry<T>();
+        qe.object = toWrap;
+        return qe;
+    }
+
+    /** The queue with the incoming objects. */
+    Queue<T> queue;
+
+    /** If we already received a EOQ event */
+    boolean closed = false;
+
+    /** Our common core */
+    private CommonCore commonCore;
+
+    /**
+     * Creates an async object which receives its data through the given queue. It will
+     * expect more events to arrive until the queue has been closed.
+     * 
+     * @param commonCore The common core. 
+     * @param queue
+     */
+    public Async(CommonCore commonCore, Queue<T> queue) {
+        this.commonCore = commonCore;
+        this.queue = queue;
+    }
+    
+    
+    /**
+     * Waits until the async has finished and returns all 
+     * remaining objects.
+     * 
+     * @since 1.0
+     * @return A {@link CoreObject} with the result.
+     */
+    public CoreObject<T> await() {
+        final Collection<T> result = $.list();
+        
+        while (true) {
+            try {
+                // Get the elements, and wait.
+                final QEntry<T> take = Async.this.queue.take();
+                if(take == EOQ) break;
+                
+                // And feed them to the listener
+                result.add(take.object);
+                
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)